### PR TITLE
Short-circuit earlier for partitioned assets with no partition key in stale resolution

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Should prevent a certain class of performance issues associated with partitioned assets on asset catalog: https://elementl-workspace.slack.com/archives/C03CCE471E0/p1693538279955379

## How I Tested These Changes

Existing test suite

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #16271** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16271" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16271?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/data_version.py [hunk 1]
```diff
@@ -403,16 +403,19 @@ def get_current_data_version(
...
     def _get_status(self, key: "AssetKeyPartitionKey") -> StaleStatus:
-        current_version = self._get_current_data_version(key=key)
-        if current_version == NULL_DATA_VERSION:
-            return StaleStatus.MISSING
-        elif self.asset_graph.is_partitioned(key.asset_key) and not key.partition_key:
-            return StaleStatus.FRESH
-        elif self.asset_graph.is_source(key.asset_key):
+        # The status loader does not support querying for the stale status of a
+        # partitioned asset without specifying a partition, so we return here.
+        if self.asset_graph.is_partitioned(key.asset_key) and not key.partition_key:
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/definitions/data_version.py:408]: just to confirm - the NULL_DATA_VERSION case is never relevant for these assets?

File: python_modules/dagster/dagster/_core/definitions/data_version.py [hunk 2]
```diff
@@ -403,16 +403,19 @@ def get_current_data_version(
...
             return StaleStatus.FRESH
         else:
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/definitions/data_version.py:410]: could leave out the else since there is a return

**Review by gibsondan — APPROVED** [review]
thanks!

File: python_modules/dagster/dagster/_core/definitions/data_version.py (hunk 1 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/data_version.py:408]: I don't believe so-- `NULL_DATA_VERSION` is relevant for individual partitions, but the status of a partitioned asset with no specified partition key is undefined, so really we can return whatever dummy value we want. If frontend is querying for this it should be ignoring the return value.

**Pull request merged**